src/db_impl.cc

```diff
@@ -1099,49 +1100,66 @@ Options TitanDBImpl::GetOptions(ColumnFamilyHandle* column_family) const {
+    const std::string& blob_run_mode_string = option_pos->second;
+    auto run_mode_mapping_pos =
+        blob_run_mode_string_map.find(blob_run_mode_string);
+    if (run_mode_mapping_pos == blob_run_mode_string_map.end()) {
+      return Status::InvalidArgument("No blob_run_mode defined for " +
+                                     blob_run_mode_string);
+    } else {
+      mutable_cf_options.blob_run_mode = run_mode_mapping_pos->second;
+      TITAN_LOG_INFO(db_options_.info_log, "[%s] Set blob_run_mode: %s",
+                     column_family->GetName().c_str(),
+                     blob_run_mode_string.c_str());
+    }
+    new_options.erase(option_pos);
+    changed = true;
+  }
+
+  option_pos = new_options.find("min_blob_size");
+  if (option_pos != new_options.end()) {
+    uint64_t min_blob_size = ParseUint64(option_pos->second);
+    mutable_cf_options.min_blob_size = min_blob_size;
+    new_options.erase(option_pos);
+    changed = true;
+  }
+  return Status::OK();
+}
+
 Status TitanDBImpl::SetOptions(
     ColumnFamilyHandle* column_family,
     const std::unordered_map<std::string, std::string>& new_options) {
   Status s;
   auto opts = new_options;
-  bool set_blob_run_mode = false;
-  TitanBlobRunMode blob_run_mode = TitanBlobRunMode::kNormal;
-  {
-    auto p = opts.find("blob_run_mode");
-    set_blob_run_mode = (p != opts.end());
-    if (set_blob_run_mode) {
-      const std::string& blob_run_mode_string = p->second;
-      auto pm = blob_run_mode_string_map.find(blob_run_mode_string);
-      if (pm == blob_run_mode_string_map.end()) {
-        return Status::InvalidArgument("No blob_run_mode defined for " +
-                                       blob_run_mode_string);
-      } else {
-        blob_run_mode = pm->second;
-        TITAN_LOG_INFO(db_options_.info_log, "[%s] Set blob_run_mode: %s",
-                       column_family->GetName().c_str(),
-                       blob_run_mode_string.c_str());
-      }
-      opts.erase(p);
-    }
-  }
+  MutableTitanCFOptions mutable_cf_options =
+      cf_info_.at(column_family->GetID()).mutable_cf_options;
+  bool titan_options_changed = false;
+  ExtractTitanCfOptions(column_family, opts, mutable_cf_options,
+                        titan_options_changed);
   if (opts.size() > 0) {
     s = db_->SetOptions(column_family, opts);
     if (!s.ok()) {
       return s;
     }
   }
-  // Make sure base db's SetOptions success before setting blob_run_mode.
-  if (set_blob_run_mode) {
+  // Make sure base db's SetOptions success before setting Titan's.
+  if (titan_options_changed) {
     uint32_t cf_id = column_family->GetID();
     {
       MutexLock l(&mutex_);
       assert(cf_info_.count(cf_id) > 0);
       TitanColumnFamilyInfo& cf_info = cf_info_[cf_id];
-      cf_info.titan_table_factory->SetBlobRunMode(blob_run_mode);
-      cf_info.mutable_cf_options.blob_run_mode = blob_run_mode;
+      cf_info.titan_table_factory->SetMutableCFOptions(mutable_cf_options);
```

Comment:
Is the mutex same as the mutex held by `SetMutableCFOptions`

Reply:
Yes, it is the db wide mutex.

Reply:
https://github.com/v01dstar/titan/blob/7085173919bd6ce906f68435c94ff8dd8ac75d91/src/db_impl.cc#L456

Reply:
Then `SetMutableCFOptions` would lock an already locked mutex?

Reply:
I use `mutext_->AssertHeld()`, right?

Reply:
TableFactory uses the same mutex, so it is `mutex->AssertHeld()`, BlobStorage uses its own mutex, so it is `MutexLock l(&(blob_storage_)mutex_)`

Reply:
Oh, I misread